src/EventStore.Core/TransactionLog/Chunks/TFChunk/TFChunk.cs

```diff
@@ -1241,5 +1221,67 @@ private struct Midpoint {
 				return string.Format("ItemIndex: {0}, LogPos: {1}", ItemIndex, LogPos);
 			}
 		}
+
+		[StructLayout(LayoutKind.Auto)]
+		private struct ReaderWorkItemPool {
+			private readonly ReaderWorkItem[] _array;
+
+			// IndexPool supports up to 64 elements with O(1) take/return time complexity.
+			// It's a thread-safe data structure with no allocations that provide predictability about
+			// the indicies: smallest available index is always preferred.
+			private IndexPool _indicies;
+
+			public ReaderWorkItemPool() {
+				_indicies = new();
+				_array = new ReaderWorkItem[IndexPool.Capacity];
+			}
+
+			// releases all available slots in the pool
+			internal int Drain(ref int referenceCount) {
+				int localReferenceCount = Interlocked.CompareExchange(ref referenceCount, 0, 0);
+				Span<int> indicies = stackalloc int[IndexPool.Capacity];
+
+				int count = _indicies.Take(indicies);
+
+				foreach (var index in indicies.Slice(0, count)) {
+					ref ReaderWorkItem slot = ref this[index];
+					slot?.Dispose();
+					slot = null;
+
+					localReferenceCount = Interlocked.Decrement(ref referenceCount);
+				}
+
+				return localReferenceCount;
+			}
+
+			internal readonly int Count => _indicies.Count;
+
+			private readonly ref ReaderWorkItem this[int index] {
+				get {
+					Debug.Assert((uint)index < (uint)_array.Length);
+
+					// skip bounds and covariance check
+					return ref Unsafe.Add(ref MemoryMarshal.GetArrayDataReference(_array), index);
+				}
+			}
+
+			internal ReaderWorkItem TryTake<T>(T arg, delegate*<T, int, ReaderWorkItem> factory) {
+				Debug.Assert(factory is not null);
+
+				return _indicies.TryTake(out int index)
+					? this[index] ??= factory(arg, index)
+					: null;
+			}
+
+			internal bool Return(ReaderWorkItem item) {
+				int index = item.PositionInPool;
+				if ((uint)index < (uint)_array.Length) {
+					_indicies.Return(index);
+					return true;
+				}
+
+				return false;
```

Comment:
```suggestion
				if (index < 0)
					return false;

				_indicies.Return(index);
				return true;
```
is there any specific reason for casting to `uint` instead of doing something like the above? (i guess it's performance related)

Reply:
> i guess it's performance related

It is. `(uint)index < (uint)_array.Length` is a fast alternative to `index > 0 && index < _array.Length`. Cast from `int` to `uint` has zero cost in assembly (it just a hint to use unsigned comparison instruction instead of signed one).

Comparison with `0` provides better readability.

Reply:
Fixed.